pkg/apis/migration/migration.go

```diff
@@ -0,0 +1,351 @@
+
+	// DeleteVolumeInfo helps delete cached mapping of volumePath to VolumeInfo and
+	// also helps delete K8s Custom Resource for specified volumePath
+	DeleteVolumeInfo(ctx context.Context, volumePath string) error
+
+	// LoadAllVolumeInfo helps cache all CR for CnsvSphereVolumeMigration
+	LoadAllVolumeInfo(ctx context.Context) error
+
+	// Register Volume takes in-tree VolumePath and helps register Volume with CNS
+	RegisterVolume(ctx context.Context, volumePath string, manager *common.Manager) (migration.CnsvSphereVolumeMigration, error)
+}
+
+// VolumeMigration holds migrated volume information and provides functionality around it.
+type volumeMigration struct {
+	// Volume Path to volumeInfo map
+	volumePathToVolumeInfo sync.Map
+	// migrationClient helps operate on CnsvSphereVolumeMigration custom resource
+	migrationClient client.Client
+}
+
+// onceForInTreeVolumeMigrationService is used for initializing the VolumeMigrationService singleton.
+var onceForVolumeMigrationService sync.Once
+
+// volumeMigrationInstance is instance of volumeMigration and implements interface for VolumeMigrationService
+var volumeMigrationInstance *volumeMigration
+
+// GetVolumeMigrationService returns the singleton VolumeMigrationService
+func GetVolumeMigrationService(ctx context.Context) (VolumeMigrationService, error) {
+	var err error
+	log := logger.GetLogger(ctx)
+	onceForVolumeMigrationService.Do(func() {
+		log.Info("Initializing volume migration service...")
+		volumeMigrationInstance = &volumeMigration{
+			volumePathToVolumeInfo: sync.Map{},
+		}
+		err = createrCnsvSphereVolumeMigrationCRD(ctx)
+		if err != nil {
+			log.Errorf("failed to create volume migration CRD. Error: %v", err)
+			return
+		}
+		config, err := k8s.GetKubeConfig(ctx)
+		if err != nil {
+			log.Errorf("failed to get kubeconfig. err: %v", err)
+			return
+		}
+		volumeMigrationInstance.migrationClient, err = k8s.NewClientForGroup(ctx, config, "cns.vmware.com")
+		if err != nil {
+			log.Errorf("failed to create migrationClient. Err: %v", err)
+			return
+		}
+		log.Info("volume migration service initialized")
+	})
+	if err != nil {
+		return nil, err
+	}
+	return volumeMigrationInstance, nil
+}
+
+// createrCnsvSphereVolumeMigrationCRD creates the CRD for CnsvSphereVolumeMigration
+func createrCnsvSphereVolumeMigrationCRD(ctx context.Context) error {
+	log := logger.GetLogger(ctx)
+	// Get a config to talk to the apiserver
+	cfg, err := config.GetConfig()
+	if err != nil {
+		log.Errorf("failed to get Kubernetes config. Err: %+v", err)
+		return err
+	}
+	apiextensionsClientSet, err := apiextensionsclient.NewForConfig(cfg)
+	if err != nil {
+		log.Errorf("failed to create Kubernetes client using config. Err: %+v", err)
+		return err
+	}
+	crdName := "cnsvspherevolumemigrations.cns.vmware.com"
+	crdKind := reflect.TypeOf(migration.CnsvSphereVolumeMigration{}).Name()
+	crd := &apiextensionsv1beta1.CustomResourceDefinition{
```

Comment:
We have a function `createCustomResourceDefinition()` that does the same thing from here onwards. Can that be moved to a common location and reused here? That will in turn reuse `waitForCustomResourceToBeEstablished()`.

I think the only change is the `Scope` of the CRD.

Reply:
added common function.

```
// createCustomResourceDefinition creates the CRD and add it into Kubernetes. If there is error,
// it will do some clean up.
func CreateCustomResourceDefinition(ctx context.Context, crdName string, crdSingular string, crdPlural string,
	crdKind string, crdGroup string, crdVersion string, crdScope apiextensionsv1beta1.ResourceScope) error {
```